File: python3/py_threads/threads_odd_even.py
"""
 * Print odd & even numbers in sync with 2 threads
 *
 * Example:
 * thread 1: 1
 * thread 2: 2
 * thread 1: 3
 * thread 2: 4
"""

from threading import Thread, Lock, Condition


class ThreadsOddEven:
    def __init__(self):
        # threads synchronization primitives
        self.mx = Lock()
        self.cv = Condition(lock=self.mx)

        # shared data between threads
        self.sd = 1

        self.max_data = 10

    # cv.wait_for() predicate objects don't take args
    def is_odd(self):
        return self.sd % 2

    def is_even(self):
        return not self.is_odd()

    def print_odd(self, tidx):
        while self.sd < self.max_data:
            # use with state for cv to acquire & release lock
            with self.cv:
                # cv.wait_for(predicate) handles spurious wake up
                self.cv.wait_for(self.is_odd)
                print(f"thread {tidx}: {self.sd}")
                self.sd += 1
                self.cv.notify()

    def print_even(self, tidx):
        while self.sd < self.max_data:
            # use with state for cv to acquire & release lock
            with self.cv:
                # cv.wait_for(predicate) handles spurious wake up
                self.cv.wait_for(self.is_even)
                print(f"thread {tidx}: {self.sd}")
                self.sd += 1
                self.cv.notify()


def main():
    th = ThreadsOddEven()
    t1 = Thread(target=th.print_odd, args=(0,))
    t2 = Thread(target=th.print_even, args=(1,))
    t1.start()
    t2.start()
    t1.join()
    t2.join()


if __name__ == "__main__":
    main()

File: cpp/cpp/threads-odd-even.cpp
/**
 * Print odd & even numbers in sync with 2 threads
 *
 * Example:
 * thread 1: 1
 * thread 2: 2
 * thread 1: 3
 * thread 2: 4
 */

#include <condition_variable>
#include <iostream>
#include <mutex>
#include <thread>

/* threads synchronization primitives */
std::mutex mx;
std::condition_variable cv;

/* shared data between threads */
static int sd = 1;

#define MAX_DATA 10

/* cv.wait() predicate objects don't take args */
static inline bool is_odd() { return sd % 2; }
static inline bool is_even() { return !is_odd(); }

void print_odd(int tidx)
{
	while (sd < MAX_DATA) {
		std::unique_lock<std::mutex> lk(mx);
		/* cv.wait(lock, predicate) handles spurious wake up */
		cv.wait(lk, is_even);
		std::cout << "thread " << tidx << ": " << sd++ << "\n";
		cv.notify_one();
	}
}

void print_even(int tidx)
{
	while (sd < MAX_DATA) {
		std::unique_lock<std::mutex> lk(mx);
		/* cv.wait(lock, predicate) handles spurious wake up */
		cv.wait(lk, is_odd);
		std::cout << "thread " << tidx << ": " << sd++ << "\n";
		cv.notify_one();
	}
}

int main(int, char **)
{
	int tidx[2] = {0, 1};
	std::thread t1(print_odd, tidx[0]);
	std::thread t2(print_even, tidx[1]);
	t1.join();
	t2.join();
	return 0;
}
